<header>
   <p  style='font-size:36px;font-family:Arial; color:#F0F0F0; background-color: #00233c; padding-left: 20pt; padding-top: 20pt;padding-bottom: 10pt; padding-right: 20pt;'>
       Mortgage Calculator chatbot using Generative AI with Vantage: Trusted AI (RAG)
  <br>
       <img id="teradata-logo" src="https://storage.googleapis.com/clearscape_analytics_demo_data/DEMO_Logo/teradata.svg" alt="Teradata" style="width: 125px; height: auto; margin-top: 20pt;">
    </p>
</header>

<p style = 'font-size:18px;font-family:Arial'><b>Introduction</b></p>

<p style = 'font-size:16px;font-family:Arial'>In the Mortgage Calculator chatbot using Generative AI demo, the combination of <b>RAG, Langchain, and LLM models</b> allows users to ask queries in layman's terms, retrieve relevant information from the Vantage tables, and generate accurate and concise answers based on the retrieved data. This integration of retrieval-based and generative-based approaches provides a powerful tool for extracting knowledge from structured sources and delivering user-friendly responses.</p>

<p style = 'font-size:16px;font-family:Arial'>In this demo we will build conversational chatbot using LangChain, a powerful library for working with LLMs like GPT-3.5, GPT-4, Mistral 7B, Mixtral 8x22B, etc. and JumpStart in ClearScape notebooks, a system is built where users can ask business questions in natural English and receive answers with data drawn from the relevant databases.</p>


<center><img src="images/header.png" alt="mortgage calc"  width=800 height=800/></center>

<br>
<p style = 'font-size:16px;font-family:Arial'>Before going any farther, let's get a better understanding of RAG, LangChain, and LLM.</p>

<ul style = 'font-size:16px;font-family:Arial'><li> <b>Retrieval-Augmented Generation (RAG):</b></li></ul>
<p style = 'font-size:16px;font-family:Arial'> &emsp;  &emsp;RAG is a framework that combines the strengths of retrieval-based and generative-based approaches in question-answering systems.It utilizes both a retrieval model and a generative model to generate high-quality answers to user queries. The retrieval model is responsible for retrieving relevant information from a knowledge source, such as a database or documents. The generative model then takes the retrieved information as input and generates concise and accurate answers in natural language.</p>

<ul style = 'font-size:16px;font-family:Arial'><li> <b>Langchain:</b></li></ul>
<p style = 'font-size:16px;font-family:Arial'> &emsp;  &emsp; LangChain is a framework that facilitates the integration and chaining of large language models with other tools and sources to build more sophisticated AI applications. LangChain does not serve its own LLMs; instead, it provides a standard way of communicating with a variety of LLMs, including those from OpenAI and HuggingFace. LangChain accelerates the development of AI applications with building blocks. We learn to leverage the following building blocks in this notebook:</p>
 
<ol style = 'font-size:16px;font-family:Arial'>
    <li> <b> LLMs</b> – LangChain's <code>llm</code> class is designed to provide a standard interface for all LLM it supports.   </li>
    <li> <b> PromptTemplate</b>  - LangChain’s <code>PromptTemplate</code> class are predefined structures for generating prompts for LLM’s. They can be reused across different LLM's.</li>
    <li> <b> Chains</b> – When we build complex AI applications, we may need to combine multiple calls to LLM’s and to other components  LangChain’s <code>chain</code> class allows us to link calls to LLM’s and components. The most common type of chaining in any LLM application is combining a prompt template with an LLM and optionally an output parser. </li>
</ol>

<ul style = 'font-size:16px;font-family:Arial'><li> <b> LLM Models (Large Language Models):</b></li></ul>
<p style = 'font-size:16px;font-family:Arial'> &emsp;  &emsp; LLM models refer to the large-scale language models that are trained on vast amounts of text data.
These models, such as <code>GPT-3.5 (Generative Pre-trained Transformer)</code>, <code>GPT-4</code> , <code>Llama 3</code> ,<code> Google's Gemini 1.5 pro</code> , <code>Anthropic Claude 3.0</code> , <code>Mistral 7B</code> ,<code>Mixtral 8x22B.</code> etc. are capable of generating human-like text responses. LLM models have been pre-trained on diverse sources of text data, enabling them to learn patterns, grammar, and context from a wide range of topics. They can be fine-tuned for specific tasks, such as question-answering, natural language understanding, and text generation.
LLM models have achieved impressive results in various natural language processing tasks and are widely used in AI applications for generating human-like text responses.</p>

<p style = 'font-size:18px;font-family:Arial'><b>What is trusted AI?</b></p>

<p style='font-size:16px;font-family:Arial'><strong>Trusted AI</strong> in the context of generative AI refers to the development and deployment of AI systems that prioritize ethical considerations, fairness, transparency, privacy, and security. Generative AI, which includes models like GPT (Generative Pre-trained Transformer), produces human-like text, images, or other data based on patterns learned from large datasets.</p>

<ul style='font-size:16px;font-family:Arial'>
  <li><strong>Ethical Considerations:</strong> Trusted generative AI operates within ethical boundaries, ensuring that the generated content aligns with societal norms and values. It avoids promoting harmful or offensive content.</li>
  <li><strong>Fairness:</strong> The AI model should produce outputs that are fair and unbiased across different demographic groups. It should not propagate stereotypes or discrimination.</li>
  <li><strong>Transparency:</strong> The inner workings of the generative AI model should be transparent and understandable to stakeholders, including developers, users, and regulatory authorities. Transparency helps build trust and facilitates accountability.</li>
  <li><strong>Privacy:</strong> Trusted generative AI respects user privacy by safeguarding sensitive information and ensuring that generated content does not compromise individuals' privacy rights. This includes anonymizing or minimizing personal data used in training.</li>
  <li><strong>Security:</strong> AI models should be designed with robust security measures to prevent malicious actors from manipulating or exploiting the system. This includes protecting against adversarial attacks and ensuring the integrity of generated content.</li>
  <li><strong>Accountability:</strong> Developers and organizations responsible for generative AI systems should be accountable for their actions and the outcomes produced by their models. This entails establishing clear lines of responsibility and mechanisms for addressing unintended consequences or ethical breaches.</li>
  <li><strong>Continual Monitoring and Improvement:</strong> Trusted AI is an ongoing process that requires continual monitoring and improvement. Developers should regularly assess the performance of their generative AI models, address emerging ethical concerns, and incorporate feedback from stakeholders to enhance trustworthiness.</li>
</ul>


<p style = 'font-size:18px;font-family:Arial'><b>Why Vantage?</b></p>


<p style = 'font-size:16px;font-family:Arial'>Chatbot, powered by Machine Learning and Artificial Intelligence, offer numerous advantages in the context of improving customer experience, particularly in the banking and financial institutions sector. Traditional methods of visiting a bank or financial institution in person and providing extensive information, which is then manually reviewed by an officer, are often time-consuming, error-prone, and rely on outdated manual processes.</p>

<p style = 'font-size:16px;font-family:Arial'>Vantage provides these same proven capabilities to search user details, proving personalized offers, helping to calculate mortgage,etc, integrated as native ClearScape Analytic functions. This allows organizations to drastically reduce human workforce by chatbot, while allowing for much more user friendly and easy interactions.</p>

<p style = 'font-size:16px;font-family:Arial'><b>Steps in the analysis:</b></p>
<ol style = 'font-size:16px;font-family:Arial'>
    <li>Configuring the environment</li>
    <li>Connect to Vantage</li>
    <li>Data Exploration</li>
    <li>Setup VectorDB</li>
    <li>Setup LLM</li>
    <li>Launch the Chatbot</li>
    <li>Cleanup</li>
</ol>

<hr style='height:2px;border:none;'>
<b style = 'font-size:20px;font-family:Arial'>1. Configuring the environment</b>

<div class="alert alert-block alert-info">
    <p style = 'font-size:16px;font-family:Arial'><i><b>Note:</b>The installation of the required libraries will take approximately <b>4 to 5 minutes</b> for the first-time installation. However, if the libraries are already installed, the execution will complete within 5 seconds.</i></p>

In [1]:
%%capture
!pip install --upgrade -r requirements.txt --quiet
# !pip install -r requirements.txt

<p style = 'font-size:16px;font-family:Arial'>
    <i>The above statements will install the required libraries to run this demo. To gain access to installed libraries after running this, restart the kernel.</i></p>

<div class="alert alert-block alert-info">
    <p style = 'font-size:16px;font-family:Arial'><i><b>Note:</b> We may need to uncomment the above statements if we run the notebooks on a platform other than ClearScape Analytics Experience that does not have the libraries installed. If we uncomment those installs, we should be sure to restart the kernel after executing those lines to bring the installed libraries into memory. The simplest way to restart the Kernel is by typing zero zero: <b>0 0</b></i></p></div>

<div class="alert alert-block alert-info">
    <p style = 'font-size:16px;font-family:Arial'><i><b>Note:</b> To ensure that our Chatbot interface reflects the latest changes, please reload the page by clicking the 'Reload' button or pressing F5 on our keyboard for <b>first-time only</b>. This will update the notebook with the latest modifications, and we'll be able to interact with the Chatbot using the new libraries.</i></p></div>

<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>1.1 Import the required libraries</b></p>

<p style = 'font-size:16px;font-family:Arial'>Here, we import the required libraries, set environment variables and environment paths (if required).</p>

In [1]:
import warnings

# Suppress warnings
warnings.filterwarnings("ignore")
display.print_sqlmr_query = False
display.suppress_vantage_runtime_warnings = True

import os
import pandas as pd

# teradata lib
from teradataml import *
display.max_rows = 5


# LLM
from langchain_community.agent_toolkits import create_sql_agent
from langchain.tools import tool
from langchain_core.messages import AIMessage
from langchain_core.prompts.chat import (
    ChatPromptTemplate,
    HumanMessagePromptTemplate,
    MessagesPlaceholder,
)
from langchain_classic.memory import ConversationSummaryBufferMemory
from langchain_core.callbacks import FileCallbackHandler
from langchain_core.prompts import (
    ChatPromptTemplate,
    HumanMessagePromptTemplate,
    MessagesPlaceholder,
)
from langchain.agents import create_agent
from langgraph.checkpoint.memory import InMemorySaver  

# from langchain.agents.format_scratchpad.openai_tools import (
#     convert_to_openai_function,
# )
from langchain_core.utils.function_calling import (
    convert_to_openai_function,
)
#from langchain.agents.output_parsers.openai_tools import OpenAIToolsAgentOutputParser
#from langchain.agents import OpenAIToolsAgentOutputParser

#from langchain.agents import AgentExecutor
from langchain_core.runnables import RunnablePassthrough
from langchain_community.utilities import SQLDatabase

# helpers
from utils.chromadb_helper import ChromaDB_VectorStore
from dotenv import load_dotenv
from litellm import completion

<hr style='height:2px;border:none;'>
<b style = 'font-size:20px;font-family:Arial'>2. Connection to Vantage and OpenAI</b>

<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>2.1 Connect to Vantage</b></p>
<p style = 'font-size:16px;font-family:Arial'>We will be prompted to provide the password. We will enter the password, press the Enter key, and then use the down arrow to go to the next cell.</p>

In [2]:
%run -i ../startup.ipynb
eng = create_context(host = 'host.docker.internal', username='demo_user', password = password)
print(eng)
execute_sql('''SET query_band='DEMO=PP_Mortgage_Calculator_Python_TrustedAI_Finetune_LLM.ipynb;' UPDATE FOR SESSION;''')

Performing setup ...
Setup complete



Enter password:  ········


... Logon successful
Connected as: teradatasql://demo_user:xxxxx@host.docker.internal/dbc
Engine(teradatasql://demo_user:***@host.docker.internal)


TeradataCursor uRowsHandle=32 bClosed=False

<p style = 'font-size:16px;font-family:Arial'>Begin running steps with Shift + Enter keys. </p>

<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>2.2 Configuring LiteLLM Proxy Environment</b></p>

<p style = 'font-size:16px;font-family:Arial'>The following cell configures the LiteLLM environment for this demo by enabling debug logging, loading credentials from a secure environment file, and routing all LLM requests through a centralized LiteLLM proxy.</p>


<ol style = 'font-size:16px;font-family:Arial'>
    <li> <b> LITELLM_LOG:</b> –  Enables detailed debug-level logging for request tracing and troubleshooting
   </li>
    <li> <b> LITELLM_API_KEY:</b>  -  Loads the LiteLLM API key securely from the environment file
</li>
    <li> <b> LITELLM_BASE_URL:</b> –  Defines the LiteLLM proxy endpoint through which all model requests are routed</li>
</ol>


In [3]:
os.environ["LITELLM_LOG"] = "DEBUG"

# Load env file
load_dotenv("/home/jovyan/JupyterLabRoot/VantageCloud_Lake/.config/.env")

os.environ["LITELLM_API_KEY"] = os.getenv("litellm_key")
litellm_key = os.environ["LITELLM_API_KEY"]

litellm_base = os.getenv('litellm_base_url')

In [4]:
os.environ["OPENAI_API_KEY"] = litellm_key
os.environ["OPENAI_API_BASE"] = litellm_base

In [5]:
if os.environ.get('LITELLM_API_KEY') and litellm_base:
    print("✓ LiteLM configured successfully")
    print(f"  Base URL: {litellm_base}")
    print(f"  API Key: {os.environ['LITELLM_API_KEY'][:10]}...")
else:
    print("✗ Error: LiteLM credentials not found in .env file")
    print("  Please check /home/jovyan/JupyterLabRoot/VantageCloud_Lake/.config/.env")

✓ LiteLM configured successfully
  Base URL: https://llmlite.ci.clearscape.teradata.com/
  API Key: sk-3RBGXDM...


<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>2.3 Getting Data for This Demo</b></p>
<p style = 'font-size:16px;font-family:Arial'>In the cell below, we are synthetically generating a few records for the <b>Customer</b> and <b>Interest</b> tables. This will allow us to test the functionality of our Chatbot with a small set of data before loading the entire dataset.For the Customer table, we are generating 10 records with randomized names, emails, and phone numbers, account details, etc.  We are also assigning a unique customer ID to each record. For the Interest table, we are generating 5 records with randomized interest rates based on minimum and maximum of credit score.  We are also assigning a unique interest ID to each record.By generating this sample data, we can quickly test the application's ability to display customer information and interests.</p>

<p style = 'font-size:16px;font-family:Arial'>In addition to this, we are loading data for <b>RealEstate, Locality, NearbyLocality, School, and ShoppingCenter</b>. We take RealEstate data from Kaggle, clean it first, and then add it to Vantage tables in the below code.</p>

<p style='font-size:16px;font-family:Arial'>We iterate through a dictionary mapping file names to primary keys, read corresponding CSV files using Pandas, and copy the data to Vantage using the teradataml <b>copy_to_sql</b> function. </p>

In [6]:
files_dict = {
    "Customer": "CustomerID",
    "Interest": "ID",
    "RealEstate": "PropertyID",
    "Locality": "LocalityID",
    "NearbyLocality": "LocalityID",
    "School": "SchoolID",
    "ShoppingCenter": "StoreID",
}

print("-" * 25, "data ingestion started", "-" * 25)
for file in files_dict:
    df = pd.read_csv(os.path.join("./data/", f"{file}.csv"))
    print("Copying data to Vantage for: ", file)
    print(f"Data information: {df.shape}")
    copy_to_sql(
        df, table_name=file, primary_index=files_dict[file], if_exists="replace"
    )

print("-" * 25, "data ingestion completed", "-" * 25)

------------------------- data ingestion started -------------------------
Copying data to Vantage for:  Customer
Data information: (10, 18)
Copying data to Vantage for:  Interest
Data information: (5, 4)
Copying data to Vantage for:  RealEstate
Data information: (285, 17)
Copying data to Vantage for:  Locality
Data information: (10, 5)
Copying data to Vantage for:  NearbyLocality
Data information: (26, 3)
Copying data to Vantage for:  School
Data information: (26, 3)
Copying data to Vantage for:  ShoppingCenter
Data information: (20, 4)
------------------------- data ingestion completed -------------------------


<hr style='height:2px;border:none;'>
<b style = 'font-size:20px;font-family:Arial'>3. Data Exploration</b>

<p style = 'font-size:16px;font-family:Arial'>The goal of the Mortgage Chatbot is to provide a user-friendly interface for individuals to quickly and easily find personalized mortgage information. The chatbot aims to streamline the mortgage process by allowing users to ask questions in natural language and receive accurate, concise answers and calculations. By integrating with Vantage, the chatbot can retrieve relevant data and perform calculations to provide users with tailored mortgage options, empowering them to make informed decisions and save time in their home buying journey. </p>

<p style = 'font-size:16px;font-family:Arial'>In our database, we have total seven tables, we can categorize it into two: </p>
    <ol style = 'font-size:16px;font-family:Arial'>
                <li> Financial tables</li>
                <li> Property tables</li>
    </ol>

<br/>
<ol style = 'font-size:16px;font-family:Arial'><li> <b>Financial tables:</b></li></ol>
<hr style='height:1px;border:none;'>

<p style = 'font-size:16px;font-family:Arial'> There are two tables: <b>customer</b> and <b>interest</b>. The customer table contains personal and financial information, while the interest table has details on credit score ranges and their corresponding interest rates.</p>

<p style = 'font-size:16px;font-family:Arial'>The data is synthetically generated, with a few rows of customer data, including name, address, email, account number, balance, income, and other relevant information.<p/>

<p style = 'font-size:16px;font-family:Arial'>Each row represents a snapshot of data captured at a specific point in time, and each column represents a different variable. The input dataset is as follows:</p>
<p style = 'font-size:16px;font-family:Arial'> 
<ul style = 'font-size:16px;font-family:Arial'>
    <li><b>Customer data</b> i.e. first-name, last-name, date of birth, address, city, account number, Balance, branch id, account status, income etc.</li>
    <li><b>Interest data</b>: min credit score, max credit score, interest rate.</li>
</ul>
</p>
<br/>

<ol style = 'font-size:16px;font-family:Arial', start=2><li> <b>Property tables:</b></li></ol>
<hr style='height:1px;border:none;'>

<p style = 'font-size:16px;font-family:Arial'>We have five tables that cover various aspects, ranging from real estate to locality details. For instance, let's take the locality of <b>Larrakeyah</b>, situated <b>9.0 km away from the airport</b>, boasting a <b>livability score of 4.3.</b> In Larrakeyah, we find <b>three schools:</b> St John's Catholic College, Larrakeyah Primary School, and Essington International School. Moreover, there are <b>two shopping centers:</b> Larrakeyah Shopping Complex and Larrakeyah Supermart.</p>

<p style = 'font-size:16px;font-family:Arial'>We see each row as a snapshot of data captured at a specific point in time, and each column represents a different variable. This is our input dataset.</p>
<p style = 'font-size:16px;font-family:Arial'> 
<ul style = 'font-size:16px;font-family:Arial'>
   <li><b>RealEstate data</b>:LocalityID, PropertyID, PropertyType, BuildingSize, LandSize, PreferredSize, OpenDate, ListingAgency, Price, Address, City, State, ZipCode, Phone, 
       BedroomCount, BathroomCount, ParkingCount</li>
    <li><b>Locality data</b>: LocalityID, LocalityName, LocalityDescription, DistanceFromAirport, LivabilityScore</li>
    <li><b>NearbyLocality data</b>: LocalityID, LocalityName, NearbyLocalityID</li>
    <li><b>School data</b>: SchoolID, SchoolName, LocalityID</li>
    <li><b>ShoppingCenter </b>: StoreID, StoreName, Type, LocalityID</li>
</ul>
</p>

<p style = 'font-size:16px;font-family:Arial'>The source data from <a href="https://www.kaggle.com/datasets/thedevastator/australian-housing-data-1000-properties-sampled">kaggle</a> is loaded in Vantage with table named <i>RealEstate</i>.</p>

<p style = 'font-size:16px;font-family:Arial'><b><i>*Please <a href="#datasetInfo1">click here</a> to scroll down to the end of the notebook for detailed column descriptions of the dataset.</i></b></p>

<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>3.1 Examine the Customer and Interest table</b></p>    
<p style = 'font-size:16px;font-family:Arial'>Let's look at the sample data in Customer table.</p>

In [7]:
tdf_customer = DataFrame("Customer")

print("Data information: \n", tdf_customer.shape)
tdf_customer

Data information: 
 (10, 18)


FirstName LastName  DateOfBirth Gender                 Address  LocalityID         City    Country                      Email    PhoneNumber AccountNumber AccountType    Balance  BranchID AccountStatus  CreditScore  Income
CustomerID                                                                                                                                                                                                                               
CID5058817      Jane    Smith   1985-09-22      F    14/14 Dashwood Place           1  Darwin City  Australia     jane.smith@example.com   987-654-3210    ACCT987654    Checking  135000.50         2        Active          680  550000
CID8039912     David   Cooper   1978-03-30      M  126/19 Kitchener Drive           2  Darwin City  Australia        david.c@example.com   111-222-4578    ACCT111222     Savings  105000.30         2        Active          700  320000
CID5058818       Sam    Smith   1985-09-22      F    14/14 Dashwood Place           4  Darwin City  Australia      sam.smith@example.com   987-654-4525    ACCT987655    Checking  135000.50         2        Active          680  550000
CID8035135     Sarah   Wilson   1980-07-10      F     24/8 Knuckey Street           5  Darwin City  Australia   sarah.wilson@example.com   321-555-9876    ACCT321555    Checking  290000.25         3        Active          800  450000
CID3058245      John      Doe   1990-04-15      M         44 Woods Street           1  Darwin City  Australia       john.doe@example.com   123-456-7890    ACCT123456     Savings  120000.00         1        Active          720  600000

<p style = 'font-size:16px;font-family:Arial'>There are 10 records in all, and there are 18 variables.</p>
<p style = 'font-size:16px;font-family:Arial'>Now, let's delve into the interest table. Here, we determine interest rates based on customers' minimum and maximum credit scores.</p>

In [8]:
tdf_interest = DataFrame("Interest")

print("Data information: \n", tdf_interest.shape)
tdf_interest

Data information: 
 (5, 4)


MinCreditScore  MaxCreditScore  InterestRate
ID                                              
1              300             579           4.8
4              740             799           3.5
2              580             669           4.3
3              670             739           3.9
5              800             850           3.0

<p style = 'font-size:16px;font-family:Arial'>Let's examine that interest rate in light of a person's credit score. Here, we may determine the interest rate for each individual consumer by examining the credit score range min and max from the interest table.</p>

In [ ]:
q = """
    SELECT c.FirstName, c.LastName, c.CreditScore, i.InterestRate 
    FROM Customer c 
    JOIN Interest i ON c.CreditScore BETWEEN i.MinCreditScore AND i.MaxCreditScore
    WHERE c.CustomerID = 'CID3058245';
"""

DataFrame.from_query(q)

<p style = 'font-size:16px;font-family:Arial'>We can observe that Customer John Doe possesses a credit score of 720, falling within the range of 670 to 739. Therefore, we will offer him a 3.9% interest rate.</p>

<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>3.2 Examine the property tables</b></p>    
<p style = 'font-size:16px;font-family:Arial'>Let's look at the sample data in <b>RealEstate</b> table first.</p>

In [ ]:
tdf_real_estate = DataFrame("RealEstate")

print("Data information: \n", tdf_real_estate.shape)
tdf_real_estate

<p style = 'font-size:16px;font-family:Arial'>We can see that we have ample details about the property here, such as property type, land size, bedroom count, contact details, price, full address, etc. These are crucial details for customers when we search for properties.</p>

<p style = 'font-size:16px;font-family:Arial'>Next, let's examine the sample data in our <b>Locality</b> table.</p>

In [ ]:
tdf_locality = DataFrame("Locality")

print("Data information: \n", tdf_locality.shape)
tdf_locality

<p style = 'font-size:16px;font-family:Arial'>In the above table, we can gain a better understanding of the locality, including its distance from the airport and livability score.</p>


<p style = 'font-size:16px;font-family:Arial'>Thirdly, let's review the sample data in our <b>NearbyLocality</b> table.</p>

In [ ]:
tdf_nearby_locality = DataFrame("NearbyLocality")

print("Data information: \n", tdf_nearby_locality.shape)
tdf_nearby_locality

<p style = 'font-size:16px;font-family:Arial'>Next, let's review the sample data in our <b>School</b> table.</p>

In [ ]:
tdf_school = DataFrame("School")

print("Data information: \n", tdf_school.shape)
tdf_school

<p style = 'font-size:16px;font-family:Arial'>Next, let's review the sample data in our <b>ShoppingCenter</b> table.</p>

In [ ]:
tdf_shopping_center = DataFrame("ShoppingCenter")

print("Data information: \n", tdf_shopping_center.shape)
tdf_shopping_center

<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>3.3 Get database schema</b></p>

<p style = 'font-size:16px;font-family:Arial'>Build a consolidated view of Table Data Catalog by combining metadata stored for the database and table.</p>

In [ ]:
main_d = {
    "Customer": tdf_customer.columns,
    "Interest": tdf_interest.columns,
    "RealEstate": tdf_real_estate.columns,
    "Locality": tdf_locality.columns,
    "NearbyLocality": tdf_nearby_locality.columns,
    "School": tdf_school.columns,
    "ShoppingCenter": tdf_shopping_center.columns,
}


def get_db_schema():
    table_dicts = []
    for k in main_d:
        table_dicts.append(
            {
                # "database_name": database,
                "table_name": k,
                "column_names": main_d[k],
            }
        )

    database_schema_string = "\n".join(
        [
            f"Table: {table['table_name']}\nColumns: {', '.join(table['column_names'])}"
            for table in table_dicts
        ]
    )

    return database_schema_string

In [ ]:
database_schema = get_db_schema()
print(database_schema)

In [ ]:
from sqlalchemy import text



def is_sql_return_results(qry):


    with eng.connect() as connection:


        results_as_dict = connection.execute(text(qry)).mappings().all()


        print(f"Total results from DB: {len(results_as_dict)}")


        return True if len(results_as_dict) > 0 else False

<hr style='height:2px;border:none;'>

<p style = 'font-size:20px;font-family:Arial'><b>4. Setup VectorDB</b></p>

<p style = 'font-size:16px;font-family:Arial'>In this step we'll build a VectorDB by passing documents, DDL and SQL queries which will helps to SQLAgent for build accurate queries. We are using Chroma vectorDB here, to read more about it please click <a href='https://docs.trychroma.com/'>here.</a> This step can be divided into 2 phases.</p>

<ul style='font-size:16px;font-family:Arial'>
    <li><b>Phase-1 Training:</b>
        <ul style='font-size:16px;font-family:Arial'>
            <li><b>Train with DDL Statements: </b>They contain information about our table names, columns, data types, and relationships in our database.</li>
            <li><b>Train with Documentation: </b>We may want to add documentation about our business terminology or definitions, which can help us, as well as the LLM, to build a better SQL.</li>
            <li><b>Train with Teradata SQL: </b>We can also add SQL queries to our training data. This is useful if we have some queries already stored in the VectorDB. So, if a user asks a similar question, the LLM can retrieve reference SQL from the VectorDB.</li></ul>
    </li>
</ul>

<ul style = 'font-size:16px;font-family:Arial'><li><b>Phase-2 Inference: </b>
<ol style='font-size:16px; font-family:Arial;'>
  <li>User asks a question.</li>
  <li>The question is converted into embeddings.</li>
  <li>These embeddings are used to search the VectorDB for similar SQL queries, tables, or documents.</li>
  <li>VectorDB returns the top 3 similar documents/SQL queries/DDL statements.</li>
  <li>The retrieved information is passed to the next step for generating the prompt.</li>
  <li>A prompt generator constructs a prompt based on the retrieved information.</li>
  <li>The constructed prompt is passed to SQLAgent or the LLM to generate the final SQL query.</li>
  <li>The final SQL query is executed on Teradata.</li>
</ol>
</li></ul>


<p style = 'font-size:16px;font-family:Arial'>The following diagram illustrates the architecture.</p>

<center><img src="images/rag_arch.png" alt="rag"  width=852 height=669/></center>

<p style = 'font-size:16px;font-family:Arial'><p style = 'font-size:16px;font-family:Arial'>In the below code will build a vectorDB and store the data under <code>chroma_vectordb</code> directive.</p></p>

In [ ]:
def train_rag(db):
    print("--" * 25, " training started ", "--" * 25)

    ## add docs
    db.add_documentation(
        documentation="Our business is dedicated to assisting users in finding the optimal properties and navigating the complexities of mortgage calculations."
    )

    # read que-sql pairs from csv
    df_sql = pd.read_csv("./data/sql_queries.csv")

    ## add SQL
    for index, row in df_sql.iterrows():
        # print(f"Question: {row['question']} \nSQL: {row['sql']}\n")
        db.add_question_sql(question=row["question"], sql=row["sql"])

    # add ddl
    db.add_ddl(ddl=database_schema)
    print("--" * 25, " training completed ", "--" * 25)


# define vectordb
chroma = ChromaDB_VectorStore()
train_rag(chroma)

<p style = 'font-size:16px;font-family:Arial'>At any time, we can inspect the training data that the package is able to reference.</p>

In [20]:
chroma.get_training_data()

id  \
0   ac8d3c72-5af4-5b8a-9303-777b7bc2246d-sql   
1   e366252a-bb15-5e1d-ad72-a2502f28041c-sql   
2   fba08a6e-6723-50e8-b6ee-b808c9616004-sql   
3   b282d065-faca-58a9-9428-0ae34426f821-sql   
4   9c355119-204c-57bb-b1e0-177ea33c7a2f-sql   
5   453a0832-951d-59de-94d1-595d74b3c8e6-sql   
6   1910077a-99b3-53ee-9739-f3045b8a0eca-sql   
7   35592eb3-9c88-5f70-817b-4bb6aabe4f31-sql   
8   b7114673-aefb-561f-be17-caa447d0ac75-sql   
9   fdb40d28-0569-5d8a-8c68-ed05f7f60748-sql   
10  f5e1c06e-0894-568a-8eda-0d54f4d3203f-sql   
11  4a109b0d-5691-59b5-853b-337d2b04cbf7-sql   
12  d1c99167-0c87-5f1b-b766-e874827e69c2-sql   
13  87df2d32-796b-5247-8cb7-138bc5fdfa97-sql   
14  17dd3be9-91ae-5254-8dcb-44624d729539-sql   
15  2bc1cf38-2f3a-5bcd-aa1e-4c0e170d5f86-sql   
16  aead94fe-79fb-5c94-a3f3-84b447a94c55-sql   
17  8649f1fd-fb74-599f-9562-3527f6a910de-sql   
18  a748a30b-daa9-5161-b248-53c24fdcdb8a-sql   
19  7c7deb35-6132-546d-921a-2c30ba81b405-sql   
20  91edf230-21ab-5488-992a-85020fd0a409-sql   
21  55b6e52a-00cb-5b31-ba7d-671ba26507ec-sql   
22  4327cc35-3006-5ad6-affe-91dab70cacf2-sql   
23  e1b8f3d7-7615-5b56-b631-acbb3584fdbb-sql   
24  6795f391-5f3c-5a9c-b964-a54d89704a38-sql   
0   81d81bb4-3895-59a4-b2e9-034b2e80dff5-ddl   
0   65761670-5638-59a0-9e11-c0b6d1db273b-doc   

                                             question  \
0   Can you suggest apartments with a land size gr...   
1   Can you suggest apartments with a land size gr...   
2   Can you suggest good properties within my budg...   
3   Is there any properties above $300,000 in Wood...   
4   Which are the best localities near the city ce...   
5   which is the best localities to live with good...   
6   What is the mean price of 3 bedroom apartments...   
7            What is the average income of customers?   
8   Is there a correlation between distance from t...   
9   What is the average price per square foot of p...   
10  How far is the property PropertyID:1350999 is ...   
11  Can you provide information about the livabili...   
12  Could you offer details regarding the surround...   
13  How many bedrooms, bathrooms, and parking spac...   
14  Can you suggest me properties within a 15-kilo...   
15  What types of properties are available in the ...   
16  Which agency handles property listings in the ...   
17  Please provide the top three most expensive pr...   
18  List out the top three least expensive propert...   
19  What is the mean income of customers in Fannie...   
20  What is the property price range in the Leanye...   
21  Which area has the most grocery stores and a g...   
22  I'm looking for a unit in Darwin City with a b...   
23  Are there any units listed by Habitat Real Est...   
24  How many properties are listed on Parap locali...   
0                                                None   
0                                                None   

                                              content training_data_type  
0   SELECT  PropertyID, PropertyType, BuildingSize...                sql  
1   SELECT PropertyID, PropertyType, BuildingSize,...                sql  
2   SELECT PropertyID, PropertyType, BuildingSize,...                sql  
3   SELECT PropertyID, PropertyType, Price, Addres...                sql  
4   SELECT TOP 3 LocalityName, LivabilityScore FRO...                sql  
5   SELECT Locality.LocalityName, COUNT(DISTINCT S...                sql  
6   SELECT AVG(Price) AS Mean_Price FROM RealEstat...                sql  
7   SELECT AVG(Income) AS AverageIncome FROM Customer                sql  
8   SELECT LocalityName, DistanceFromAirport, AVG(...                sql  
9   SELECT AVG(Price / BuildingSize) AS AvgPricePe...                sql  
10  SELECT RealEstate.PropertyID, RealEstate.Addre...                sql  
11  SELECT LocalityName, LivabilityScore FROM Loca...                sql  
12  SELECT LocalityName FROM Locality WHERE Locali...                sql  
13  SELECT BedroomCount, BathroomCount, ParkingCou

<p style = 'font-size:16px;font-family:Arial'>We have the capability to remove training data if there's obsolete or incorrect information by passing the <code>xxx-id</code></p>

In [21]:
# chroma.remove_training_data(id="xx-sql")

<p style = 'font-size:16px;font-family:Arial'>We utilize this method to generate a prompt for the LLM (Large Language Model) to generate accurate SQL queries by passing user's question.</p>

In [22]:
def generate_prompt_from_vectordb(question):
    question_sql_list = chroma.get_similar_question_sql(question)
    ddl_list = chroma.get_related_ddl(question)
    doc_list = chroma.get_related_documentation(question)

    initial_prompt = None
    return chroma.get_sql_prompt(
        initial_prompt=initial_prompt,
        question=question,
        question_sql_list=question_sql_list,
        ddl_list=ddl_list,
        doc_list=doc_list,
    )

<p style = 'font-size:16px;font-family:Arial'>To take a look at how the prompt looks, simply uncomment the code below and modify the question if desired.</p>

<hr style='height:2px;border:none;'>
<b style = 'font-size:20px;font-family:Arial'>5. Setup LLM </b>
<hr style='height:1px;border:none;'>
<p style = 'font-size:18px;font-family:Arial'><b>5.1 Connect to database using SQLAlchemy and Initialize the Large Language Model</b></p>    

<p style = 'font-size:16px;font-family:Arial'>Under the hood, LangChain uses SQLAlchemy to connect to SQL database. The SQLDatabaseToolkit can therefore be used with any SQL dialect supported by SQLAlchemy, such as Teradata Vantage, MS SQL, MySQL, MariaDB, PostgreSQL, Oracle SQL, and SQLite. Please refer to the <a href="https://docs.sqlalchemy.org/en/20/"> SQLAlchemy documentation</a> for more information about requirements for connecting to your database. The SQLDatabaseToolkit builds off of SQLDatabaseChain and is designed to answer more general questions about a database, as well as recover from errors.</p>


<div class="alert alert-block alert-info">
    <p style = 'font-size:16px;font-family:Arial'><i><b>Important Note:</b> The code below establishes a database connection for data sources and Large Language Models. Please note that the solution will only work if the database connection for your sources is defined in the cell below. In addition to this, to use OpenAI models, we have to set OpenAI API key to environment variable.</i></p></div>

In [24]:
# define DB
db = SQLDatabase(eng)

from langchain_openai import ChatOpenAI

# # OpenAI API
# os.environ["OPENAI_API_KEY"] = litellm_key

llm = ChatOpenAI(model="openai/gpt-4o", temperature=0.7)

In [25]:
# from litellm import completion
# import os

# ## set ENV variables
# os.environ["OPENAI_API_KEY"] = litellm_key

# response = completion(
#   model="openai/gpt-4o",
#   messages=[{ "content": "Hello, how are you?","role": "user"}]
# )

<p style = 'font-size:16px;font-family:Arial'>In this demo we will be using OpenAI’s <code>gpt-3.5-turbo-0125</code> model as LLM. To view list of available models on OpenAI <a href='https://platform.openai.com/docs/models'>click here</a></p>

<p style = 'font-size:16px;font-family:Arial'>In OpenAI's language models, the <b>temperature</b> parameter controls the randomness of the generated text. It affects the diversity and creativity of the model's responses. It is always a number between 0 and 1. A temperature of 0 means the responses will be very straightforward, almost deterministic (meaning we almost always get the same response to a given prompt). A temperature of 1 means the responses can vary wildly.</p>


<p style = 'font-size:16px;font-family:Arial'>A higher temperature value, such as 1.0, increases the randomness and diversity of t|he generated output. This can lead to more varied and surprising responses, but it may also result in less coherence and occasional nonsensical outputs. A higher temperature means that the model might select a word with slightly lower probability, leading to more variation, randomness and creativity. A very high temperature therefore increases the risk of <b>hallucination</b>, meaning that the model starts selecting words that will make no sense or be off the topic.</p>

<p style = 'font-size:16px;font-family:Arial'>On the other hand, a lower temperature value, such as 0.2 or below, reduces randomness and makes the model's output more focused and deterministic. The generated text is likely to be more conservative, sticking closely to patterns observed in the training data. A temperature of 0 means roughly that the model will always select the highest probability word.</p>

<p style = 'font-size:16px;font-family:Arial'>Choosing an appropriate temperature value depends on the desired output. Higher temperatures can be useful for creative tasks or brainstorming, while lower temperatures are preferred when we need more control over the output, such as when generating specific responses or following a particular style.</p>

<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>5.2 Setup SQLAgent</b></p> 

<p style = 'font-size:16px;font-family:Arial'>Before jumping into the chatbot, let us first understand what is an agent and why it might be preferred over a simple SQLChain.</p>

<p style = 'font-size:16px;font-family:Arial'>An agent is a component that has access to a suite of tools, including a Large Language Model (LLM). Its distinguishing characteristic lies in its ability to make informed decisions based on user input, utilizing the appropriate tools until it achieves a satisfactory answer. For example in the context of text-to-SQL, the LangChain SQLAgent will not give up if there is an error in executing the generated SQL. Instead, it will attempt to recover by interpreting the error in a subsequent LLM call and rectify the issue. Therefore, in theory, SQLAgent should outperform SQLChain in productivity and accuracy.</p>
    
<p style = 'font-size:16px;font-family:Arial'>We can think of agents as enabling tools for LLMs. Like how a human would use a calculator for maths or perform a Google search for information — agents allow an LLM to do the same thing.</p>

In [26]:
from typing import Any, Dict
import uuid

from langchain_core.messages import (
    AIMessage,
    BaseMessage,
    HumanMessage,
    SystemMessage,
    ToolMessage,
)

#from langchain.callbacks.base import BaseCallbackHandler
from langchain_core.callbacks import BaseCallbackHandler

from langchain_core.agents import AgentAction
import ast


# First, define custom callback handler implementations
class ToolCallbackHandler(BaseCallbackHandler):
    def __init__(self, sql=None):
        if sql is None:
            sql = {}

    def on_tool_start(
        self, serialized: Dict[str, Any], input_str: str, **kwargs: Any
    ) -> Any:
        name = serialized.get("name")
        print(f"*** on_tool_start: *** {name}")
        if name in ["sql_db_query", "sql_db_query_checker"]:
            _sql = ast.literal_eval(input_str)["query"]
            print("*** _sql: *** ", _sql)
            if is_sql_return_results(_sql):
                self.sql = _sql
                print(f"sql assigned: {self.sql}")
            else:
                print("sql not assigned")
                self.sql = {}
            return self.sql

    def on_llm_end(self, output: str, **kwargs: Any) -> Any:
        txt = output.generations[0][0].text
        if txt != "":
            print("--- on_llm_end ---")
            # print('TXT: \n\n', txt, '\n\n')
            resp_dict = ast.literal_eval(output.generations[0][0].text)
            _sql = resp_dict.get("SQL", "Key not found")
            _error = resp_dict.get("error", "Key not found")
            print(f"_sql: {_sql}")
            print(f"_error: {_error}")

            # set SQL
            if is_sql_return_results(_sql):
                self.sql2 = _sql
                print(f"sql2 assigned: {self.sql2}")
            else:
                print("sql2 not assigned")
                self.sql2 = {}
            return self.sql2


# define handler
tool_handler = ToolCallbackHandler()


# define SQL Agent
@tool
def generate_sql(input: str) -> str:
    """Given an input question, first create a syntactically correct Teradata-style query to execute, then look at the results of the query and return the answer."""
    generated_prompt = generate_prompt_from_vectordb(input)

    generated_prompt = (
        generated_prompt
        + """Alway return Final output in json format.
                Final output:
                    - SQL:
                    - Answer:"""
    )
    # print(f"\n\n generated_prompt: \n\n{generated_prompt} \n\n")

    messages = [
        HumanMessagePromptTemplate.from_template("{input}"),
        AIMessage(content=generated_prompt),
        MessagesPlaceholder(variable_name="agent_scratchpad"),
    ]

    prompt = ChatPromptTemplate.from_messages(messages)
    agent_executor = create_sql_agent(
        llm,
        db=db,
        agent_type="openai-tools",
        verbose=True,
        prompt=prompt,
        max_iterations=10,
        max_execution_time=20,
        handle_parsing_errors=True,
        return_intermediate_steps=True,
        handle_sql_errors=True,
        max_tokens=4000,
    )

    return agent_executor.run(input, callbacks=[tool_handler])

<p style = 'font-size:16px;font-family:Arial'>In the code above, we're initializing SQLAgent by passing in LLM, SQLDatabaseToolkit, database, and agent type as OpenAI tools. Agents use an LLM to determine the best course of action and execute it. An action can either be using a tool and observing its output, or responding to the user. <a href='https://python.langchain.com/docs/modules/agents/agent_types/'>Here</a> are the agents available in LangChain.</p>

<p style = 'font-size:16px;font-family:Arial'>OpenAI tools in LangChain refer to the integration of OpenAI's language models with external tools and services to enhance their capabilities and provide more accurate and informative responses. LangChain is a framework that enables the development of applications powered by language models, and it supports the use of OpenAI models as the language understanding component.</p>
<p style = 'font-size:16px;font-family:Arial'>The OpenAI tools in LangChain are designed to overcome the limitations of language models, such as the lack of recency and specificity in their training data, by allowing them to access external knowledge bases and APIs dynamically. This enables the models to answer questions with context directly from search engines, APIs, or internal databases, and to perform intermediate steps to gather relevant information. Please refer this <a href='https://python.langchain.com/v0.1/docs/modules/agents/agent_types/openai_tools/'>openai-tools</a> for more information.</p>


<hr style='height:2px;border:none;'>
<b style = 'font-size:20px;font-family:Arial'>6. Launch the Chatbot</b>

<p style = 'font-size:16px;font-family:Arial'>In this demo we are using ChatOpenAI model with Memory. This advanced technology allows us to store and recall conversations, enabling our chatbot to provide more personalized and informed responses.As a mortgage advisor, our chatbot is trained to assist with a wide range of loan-related inquiries. Whether you're looking to purchase a new home, refinance an existing loan, or simply have questions about the mortgage process, our chatbot is here to help.To begin, we'll set the prompt for our chatbot to work as a mortgage advisor. This will enable it to provide tailored advice and guidance based on your unique needs and circumstances.</p>

<div class="alert alert-block alert-info"> <p style = 'font-size:16px;font-family:Arial'><i><b>Note:</b>Please note that our chatbot is specifically designed to address questions related to mortgage loans. If we have a question outside of this scope, we kindly ask that we refrain from asking it. This will help us provide ourselves with the most accurate and efficient assistance possible.</i></p> </div>


In [27]:
from datetime import datetime

# default customer
CustomerID = "CID3058245"

# define tools
tools = [generate_sql]
from langchain.tools import tool

@tool
def generate_sql(query: str) -> str:
    """Use this to fetch answers from database."""
    return run_your_sql_logic(query)

llm_with_tools = llm.bind_tools(tools)
# memory = ConversationSummaryBufferMemory(
#     return_messages=True,
#     memory_key="chat_history",
#     llm=llm,
# )

checkpointer = InMemorySaver()

# For writing logs
from loguru import logger

now = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
logfile = f"./logs/rag_sql_agent_output_{now}.log"
logger.add(
    logfile,
    colorize=True,
    enqueue=True,
    level="DEBUG",
    backtrace=True,
    diagnose=True,
    format="{time} {level} {message}",
)
file_handler = FileCallbackHandler(logfile)
config = {"callbacks": [file_handler]}

# main prompt
prompt = ChatPromptTemplate.from_messages(
    [
        (
            "system",
            f"""You'll be working as property advisor and mortgage advisor and I will play the role of the user.
            
            * General Instructions for Chatbot:
            1) Start the chat with greeting and ask the questions to user as mentioned below. Greet user by starting the message: 
              ## Welcome to Property and Mortgage Chatbot! \n
              I'm here to provide you with expert advice and answers. \n
              ### Let's get started! \n
              Could you please share your property preferences with me?
            2) Ask questions sequentially - pausing between each question to wait for a response before proceeding to the next.
            3) Most important: Do not ask all the questions together, ask one by one.
            4) Whenever possible, give the answer in bulleted points and title in markup. like ##
            5) Consider current user: {CustomerID}
            6) If the user fails to provide a response or says 'I don't know' for a question, automatically call the 'generate_sql' function to retrieve the answer from the database.
            7) Most important: If 'generate_sql' function fail to give the answer, then just say "Sorry, I am unable to get the answer." Do not make any answer by yourself.
            
            * Instructions to follow when working as a property advisor:
            1) As a property advisor first ask a series of questions related to property like preferred locality, number of bedrooms, etc.  
            2) As a property advisor, Always end your response with the next Question. Do not ask same question more than one time.
            3) As a property advisor, you have direct access of property data includes tables like RealEstate, Locality, NearbyLocality,  ShoppingCenter, School. 
            Only Suggest properties that exist in these tables.
            4) Always give property options with details like Property ID, Property Type, Building Size, Price, Address, Bedroom Count.
            5) Most important: Once you suggest the property, ask user Do you like this or shall I suggest more? Once the user select the property, then only perform a role of mortgage advisor.
            
            * Instructions to follow when working as a mortgage advisor:
            1) Most important: Once user select the property then only ask for mortgage related questions like monthly income, credit score, bank balance etc.
            2) As a mortgage advisor for the bank you have direct access to the banks data for user. The banks data includes tables like Customer and Interest
            3) Once user select a property, then only ask about mortgage related questions. You can ask question like below one by one:
             -  question 1: Income: What is your total annual income, including any additional sources of income?
             -  question 2: What is your employment status?
             -  question 3: What's your credit score? I can help you find out which interest rates you qualify for.
            4)  Do not call unnecessarily the 'generate_sql' function until it is really needed when working as mortgage advisor.
            5) Do not call unnecessarily the 'generate_sql' function until it is really needed when working as mortgage advisor.
            6) At the end of above 3 questions provide a succinct summary paragraph of these details.
            
            
            * Remember the Instructions for final output:
            1) Provide a succinct summary paragraph of these details as a response to the client but don't show the method of calculation. 
            Follow with a formatted table of monthly amounts (gross income, tax obligation, mortgage obligation, expenses, remaining disposable income).
            Also provide best property options based on user's data.
            2) Provide an opinion as to whether the mortgage is affordable and what the maximum borrowing amount could be
            3) Most important: If 'generate_sql' function fail to give the answer, then just say "Sorry, I am unable to get the answer." Do not make any answer by yourself.
            4) Most Important: Give me reasoning as well as answers for this given questions. 
                Instructions for Reasoning:
                    - Give me Reasoning in details.
                    - Only one sentence reasoning would be good.
                    - Be transparent and Unbiased so user can trust on answer.
            5) Most important: Reasoning is mandatory in the final summary.""",
        ),
        MessagesPlaceholder(variable_name="chat_history"),
        ("user", "{input}"),
        MessagesPlaceholder(variable_name="agent_scratchpad"),
    ]
)


# define agent
# agent = (
#     RunnablePassthrough.assign(
#         agent_scratchpad=lambda x: convert_to_openai_function(
#             x["intermediate_steps"]
#         )
#     )
#     | prompt
#     | llm_with_tools
#     | OpenAIToolsAgentOutputParser()
# )

# # define Agent executor
# agent_executor = AgentExecutor(
#     agent=agent,
#     tools=tools,
#     verbose=True,
#     memory=memory,
#     max_iterations=10,
#     handle_parsing_errors=True,
#     max_execution_time=20,
# )

systemPrompt = """You'll be working as property advisor and mortgage advisor and I will play the role of the user.
            
            * General Instructions for Chatbot:
            1) Start the chat with greeting and ask the questions to user as mentioned below. Greet user by starting the message: 
              ## Welcome to Property and Mortgage Chatbot! \n
              I'm here to provide you with expert advice and answers. \n
              ### Let's get started! \n
              Could you please share your property preferences with me?
            2) Ask questions sequentially - pausing between each question to wait for a response before proceeding to the next.
            3) Most important: Do not ask all the questions together, ask one by one.
            4) Whenever possible, give the answer in bulleted points and title in markup. like ##
            5) Consider current user: {CustomerID}
            6) If the user fails to provide a response or says 'I don't know' for a question, automatically call the 'generate_sql' function to retrieve the answer from the database.
            7) Most important: If 'generate_sql' function fail to give the answer, then just say "Sorry, I am unable to get the answer." Do not make any answer by yourself.
            
            * Instructions to follow when working as a property advisor:
            1) As a property advisor first ask a series of questions related to property like preferred locality, number of bedrooms, etc.  
            2) As a property advisor, Always end your response with the next Question. Do not ask same question more than one time.
            3) As a property advisor, you have direct access of property data includes tables like RealEstate, Locality, NearbyLocality,  ShoppingCenter, School. 
            Only Suggest properties that exist in these tables.
            4) Always give property options with details like Property ID, Property Type, Building Size, Price, Address, Bedroom Count.
            5) Most important: Once you suggest the property, ask user Do you like this or shall I suggest more? Once the user select the property, then only perform a role of mortgage advisor.
            
            * Instructions to follow when working as a mortgage advisor:
            1) Most important: Once user select the property then only ask for mortgage related questions like monthly income, credit score, bank balance etc.
            2) As a mortgage advisor for the bank you have direct access to the banks data for user. The banks data includes tables like Customer and Interest
            3) Once user select a property, then only ask about mortgage related questions. You can ask question like below one by one:
             -  question 1: Income: What is your total annual income, including any additional sources of income?
             -  question 2: What is your employment status?
             -  question 3: What's your credit score? I can help you find out which interest rates you qualify for.
            4)  Do not call unnecessarily the 'generate_sql' function until it is really needed when working as mortgage advisor.
            5) Do not call unnecessarily the 'generate_sql' function until it is really needed when working as mortgage advisor.
            6) At the end of above 3 questions provide a succinct summary paragraph of these details.
            
            
            * Remember the Instructions for final output:
            1) Provide a succinct summary paragraph of these details as a response to the client but don't show the method of calculation. 
            Follow with a formatted table of monthly amounts (gross income, tax obligation, mortgage obligation, expenses, remaining disposable income).
            Also provide best property options based on user's data.
            2) Provide an opinion as to whether the mortgage is affordable and what the maximum borrowing amount could be
            3) Most important: If 'generate_sql' function fail to give the answer, then just say "Sorry, I am unable to get the answer." Do not make any answer by yourself.
            4) Most Important: Give me reasoning as well as answers for this given questions. 
                Instructions for Reasoning:
                    - Give me Reasoning in details.
                    - Only one sentence reasoning would be good.
                    - Be transparent and Unbiased so user can trust on answer.
            5) Most important: Reasoning is mandatory in the final summary."""

agent = create_agent(
    model=llm,
    tools=[generate_sql],
    system_prompt=systemPrompt,
   # checkpointer=checkpointer,
)

# agent = createAgent({
#   model: ,
#   systemPrompt: systemPrompt,
#   tools: tools,
#   responseFormat,
#   checkpointer,
# });



In [28]:
def get_properties(response):
    prop_ids = re.findall(r"Property ID:\*\* (\d+)", response)
    if len(prop_ids) == 0:
        prop_ids = re.findall(r"PropertyID:\*\* (\d+)", response)
    if len(prop_ids) == 0:
        prop_ids = re.findall(r"Property ID: (\d+)", response)
    return prop_ids


def validate_property_ids(property_ids):
    property_ids = ",".join(property_ids)

    q = f"""SELECT PropertyID FROM demo_user.RealEstate WHERE PropertyID in ({property_ids})"""
    temp_df1 = DataFrame.from_query(q)
    logger.info(f"temp_df1 shape: {temp_df1.shape}")

    return True if len(property_ids.split(",")) == temp_df1.shape[0] else False


def find_partial_match(arr, target, min_match=3):
    for element in arr:
        count = 0
        for word in element.split():
            if word.lower() in target.lower():
                count += 1
        if count >= min_match:
            return True
    return False


questions = [
    "What is your total annual income, including any additional sources of income?",
    "What is your employment status?",
    "What's your credit score?",
]


def validate_response(response_output):
    apl_msg = """We apologize for the inconvenience. There could be various reasons for this issue, including:
    - It seems we couldn't find an answer, which might be due to limited data and table information.
    - To enhance our suggestions, could you please provide additional criteria or try asking the same question again later?
    - Technical issues might be causing this delay.
    - Thank you for your cooperation. Please try again later."""

    if "Welcome to Property and Mortgage Chatbot" in response_output:
        logger.info(" Entered into greet")
        if hasattr(tool_handler, "sql"):
            tool_handler.sql = {}
        return True, response_output
    elif "Agent stopped due to max iterations" in response_output:
        logger.info(" Agent stopped due to max iterations ")
        if hasattr(tool_handler, "sql"):
            logger.info(f" tool_handler {tool_handler.sql} \n resetting now")
            tool_handler.sql = {}
        logger.info(
            " Agent stopped due to max iterations, exiting with apologies msg... "
        )
        return False, apl_msg
    elif "Sorry, I am unable to get the answer" not in response_output:
        logger.info(
            " looks valid response, validating properties and SQL", level="DEBUG"
        )
        # get prop ids from response
        property_ids = get_properties(response_output)
        logger.info(f"found property_ids: {property_ids}")

        # first validate properties if retrieved property_ids
        if len(property_ids) > 0 and validate_property_ids(property_ids):
            logger.info(" property validated ")

            if hasattr(tool_handler, "sql"):
                logger.info(f" tool_handler SQL: {tool_handler.sql} \n resetting now")
                tool_handler.sql = {}
            logger.info(" property validation done.. returning the response..")
            return True, response_output
        elif (
            hasattr(tool_handler, "sql")
            and len(tool_handler.sql) > 0
            and is_sql_return_results(tool_handler.sql)
        ):  # if not a property related question
            logger.info(" there are valid SQL but not valid property ")

            if hasattr(tool_handler, "sql"):
                logger.info(f" tool_handler SQL: {tool_handler.sql} \n resetting now")
                tool_handler.sql = {}

            logger.info(" SQL is valid, returning the response..")
            return True, response_output
        else:
            logger.info(
                " there are NO valid SQL and no valid property, might be working as mortgage advisor... "
            )

            if hasattr(tool_handler, "sql"):
                logger.info(f" tool_handler SQL: {tool_handler.sql} \n resetting now")
                tool_handler.sql = {}

            if find_partial_match(questions, response_output):
                logger.info(
                    " Found at least 3 matching words! seems it working as mortgage advisor.. returning the response.."
                )
                return False, response_output

            logger.info(
                " there are NO valid SQL and no valid property, not working as mortgage advisor, exiting with apologies msg... "
            )
            return False, apl_msg
    else:
        if hasattr(tool_handler, "sql"):
            logger.info(f" tool_handler SQL: {tool_handler.sql} \ resetting now")
            tool_handler.sql = {}

        logger.info("It seems wrong response, exiting with apologies msg... ")
        return False, apl_msg

<div class="alert alert-block alert-info">
    <p style = 'font-size:16px;font-family:Arial'><i><b>Note:</b>Chatbot is accessing multiple components, including databases and LLMs. This may cause a brief delay in responses. Your patience is appreciated.</i></p>
</div>

<hr style='height:1px;border:none;'>
<p style = 'font-size:18px;font-family:Arial'><b>6.1 User Guide: Property and Mortgage Chatbot</b></p>

<p style = 'font-size:16px;font-family:Arial'>Welcome to our Property and Mortgage Chatbot We're here to help ourselves effectively interact with our chatbot, providing assistance in both property selection and mortgage advice. Whether we're looking for our dream home or seeking mortgage options, we're here to guide ourselves every step of the way.</p>

<p style = 'font-size:16px;font-family:Arial'><b>Getting Started:</b></p>
<ol style = 'font-size:16px;font-family:Arial'>
  <li>Initiate the chatbot by sending us a greeting message. Simply say "Hello" or any other friendly greeting to start the conversation.</li>
</ol>


<p style = 'font-size:16px;font-family:Arial'><b>Property Advisor:</b></p>

<ol style = 'font-size:16px;font-family:Arial'>
  <li><strong>Property Search:</strong> Ask us questions about properties, localities, amenities, etc. For example:
    <ul style = 'font-size:16px;font-family:Arial'>
            <li><i>What is the mean price of 3 bedroom apartments in Larrakeyah area?</i></li>
            <li><i>Can you provide information about the livability of the Bayview area?</i></li>
            <li><i>Which agency handles property listings in the Woolner area?</i></li>
            <li><i>Can you suggest apartments with a land size greater than 300 square meters and featuring at least two bedrooms?</i> </li>
        </ul></li>
  <li><strong>Select Property:</strong> Once we find a property of interest, note down its Property ID.</li>
  <li><strong>Switch to Mortgage Advisor:</strong> To proceed to the mortgage advisory phase, provide type text with the Property ID of the selected property. For example: <i>I really like this property with ID 1351007</i></li>
</ol>

<p style = 'font-size:16px;font-family:Arial'><b>Mortgage Advisor:</b></p>

<ol  style = 'font-size:16px;font-family:Arial'>
  <li><strong>Provide Details:</strong> We'll ask for specific details such as your annual income, employment status, etc.</li>
  <li><strong>Summary:</strong> Based on the information we provide, we'll generate a summary of the property and mortgage details.</li>
  <li><strong>Affordability Calculation:</strong> We will receive information on the maximum house price we can afford based on our income and other relevant factors.</li>
</ol>

<p style = 'font-size:16px;font-family:Arial'><b>Tips for Effective Interaction:</b></p>

<ul  style = 'font-size:16px;font-family:Arial'>
  <li>Be specific and clear with your queries to receive accurate responses.</li>
  <li>Provide accurate information when prompted, as it will help us provide personalized assistance.</li>
  <li>We note down the Property ID of any property we are interested in to seamlessly transition to the mortgage advisory phase.</li>
</ul>


<p style = 'font-size:16px;font-family:Arial'>Our Property and Mortgage Chatbot is designed to simplify our property search and mortgage process. By seamlessly transitioning from property advisor to mortgage advisor, we provide comprehensive assistance tailored to our needs. Start our journey now by sending a greeting message and let us guide ourselves towards our dream home and a suitable mortgage option!</p>


<div class="alert alert-block alert-info">
    <p style = 'font-size:16px;font-family:Arial'><i><b>Sample summary: </b>An example of the tentative output the chatbot will provide at the end is shown below.</i></p>
</div>

<pre style='border: 1px; border-color:#00233C; border-style: outset; padding: 15px'>Based on the data, the credit score of the user with CID3058245 is 720.

Summary: The user has a monthly income of $5000, is employed full-time, and has a credit score of 720.

<b> Mortgage Details Summary:</b>
- <b>Income:</b> $5000
- <b>Employment Status:</b> Full-time
- <b>Credit Score:</b> 720

<b>Monthly Amounts:</b>
| Gross Income | Tax Obligation | Mortgage Obligation | Expenses | Remaining Disposable Income |
|--------------|----------------|---------------------|----------|-----------------------------|
| $5000        | $750           | $1500               | $1000    | $1750                       |

<b>Best Property Options based on User's Data:</b>
- <b>Property ID: 1351007</b>
  - Property Type: Apartment
  - Building Size: 120 sqm
  - Price: $450,000
  - Address: 2/1 Brolga Street, Rapid Creek
  - Bedroom Count: 2
  - Bathroom Count: 2
  - Parking Count: 1

<b>Affordability Opinion:</b> The mortgage is affordable with a remaining disposable income of $1750.

<b>Maximum Borrowing Amount:</b> $150,000

<b>Reasoning:</b> The user's income of $5000, employment status, and credit score of 720 make the mortgage affordable with a comfortable remaining disposable income.</pre>

<hr style='height:1px;border:none;'>
<p style = 'font-size:18px;font-family:Arial'><b>6.2 Open the Chabot</b></p>

<div class="alert alert-block alert-success">
    <p style = 'font-size:16px;font-family:Arial'><i>It's time to chat with the bot! Let's get started with our conversation.</i></p>
</div>


<div class="alert alert-block alert-info">
    <p style = 'font-size:16px;font-family:Arial'><b>Important Note: </b> This chatbot is designed to minimize hallucinations. If we detect an inaccurate response or experience technical issues, we will notify you with "We apologize for the inconvenience." If you encounter this error, <b>please retry your question to receive a corrected answer.</b></p></div>

In [29]:
import panel as pn

pn.extension(design="material")

# clear the memory
#memory.clear()


def callback(contents, user, instance):
    response = agent.invoke(
        {"messages": [{"role": "user", "content": contents}]},
        config=config
    )

    # Depending on your agent type:
    final_output = response.get("structured_response") or response["messages"][-1].content

    is_valid_response, final_res = validate_response(final_output)

    if (
        len(contents) > 30
        and is_valid_response
        and hasattr(tool_handler, "sql")
        and len(tool_handler.sql) > 0
    ):
        chroma.add_question_sql(question=contents, sql=tool_handler.sql)

    return final_res



chat = pn.chat.ChatInterface(
    callback=callback,
    show_rerun=False,
    show_undo=False,
    show_clear=False,
    width=800,
    height=400,
)


#chat.send("👋 Hello! I'm your AI assistant. How can I help you today?", user="Assistant")

chat.servable()

ChatInterface(_button_data={'send': _ChatButtonData(i...}, _buttons={'send': Button(align='cen...}, _input_container=Row, _input_layout=Row, _placeholder=ChatMessage, _widgets={'ChatAreaInput': ChatArea...}, callback=<function callback a..., design=<class 'panel.theme.materi..., height=400, show_button_name=True, show_clear=False, show_rerun=False, show_undo=False, sizing_mode='fixed', widgets=[ChatAreaInput(css_classes...], width=800)

<div class="alert alert-block alert-info">
    <p style='font-size:16px;font-family:Arial'>If the chatbot didn't work when we pressed ENTER, on our first time using this demo on our environment, did we use F5 to reload the site? See instructions at the top of the notebook. If we asked a question and got no response after a few minutes, it is possible that we will need to type 0 0 to restart the kernel and re-run the demo. Questions outside the model seem to confuse the chatbot.</p>

<hr style='height:2px;border:none;'>
<b style = 'font-size:20px;font-family:Arial'>7 Cleanup</b>
<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'><b>7.1 Work Tables</b></p>
<p style = 'font-size:16px;font-family:Arial'>Cleanup work tables to prevent errors next time.</p>

<hr style='height:1px;border:none;'>

<p style = 'font-size:18px;font-family:Arial'> <b>7.2 Databases and Tables </b></p>
<p style = 'font-size:16px;font-family:Arial'>The following code will clean up tables and databases created above.</p>

In [ ]:
for t in [
    "Customer",
    "Interest",
    "RealEstate",
    "Locality",
    "NearbyLocality",
    "School",
    "ShoppingCenter",
]:

    try:
        db_drop_table(t)

    except:
        pass

In [ ]:
remove_context()

<b id="datasetInfo1" style = 'font-size:20px;font-family:Arial'>Dataset:</b>

**Customer Table**
- `CustomerID`: unique row customer id
- `FirstName`: customer first name
- `LastName`: customer last name
- `DateOfBirth`: customer birth date
- `Gender` : customer gender (categorical: "M","F",)
- `Address` : customer full address
- `City`: city of customer (categorical: 'Almond','Geneva')
- `LocalityID`: unique row locality id (Numeric)
- `Country`: country of customer (categorical: 'USA')
- `Email`: customer email
- `PhoneNumber`: customer phone number
- `AccountNumber`: customer account number
- `AccountType`: customer account type
- `AccountStatus`: customer account status (categorical: 'Active','Inactive')
- `Balance`: account balance, in dollar (numeric)
- `BranchID`: branch id
- `CreditScore`: customer credit score
- `Income`: customer's monthly income, in dollar (numeric)

**Interest Table**
- `ID`: unique row id
- `MinCreditScore`: min credit score
- `MaxCreditScore`: max credit score
- `InterestRate`: rate of interest

**RealEstate**
- `LocalityID`: unique row locality id (Numeric)
- `PropertyID`: unique row property id (Numeric)
- `PropertyType`: type of property being listed (categorical: 'Apartment','House', 'Unit')
- `BuildingSize`: size of the property's building, in square meters. (Numeric)
- `LandSize`: size of the property's land, in square meters. (Numeric)
- `PreferredSize`: preferred size of the property, in square meters. (Numeric)
- `OpenDate`: date that the property was first listed for sale. (Date)
- `ListingAgency`: agency that is listing the property
- `Price`: listing price of the property
- `Address`: property's address
- `City`: city that the property is located in
- `State`: state that the property is located in
- `ZipCode`: zip code that the property is located in
- `Phone`: listing agent's phone number
- `BedroomCount`: number of bedrooms in the property
- `BathroomCount`: number of bathrooms in the property
- `ParkingCount`: number of parking spaces in the property

 **Locality data**
 - `LocalityID`: unique row locality id (Numeric)
 - `LocalityName`: locality name
 - `LocalityDescription`: locality description
 - `DistanceFromAirport`: distance from airport to locality (Numeric)
 - `LivabilityScore`: locality livability score (Numeric)
 
**NearbyLocality data**
- `LocalityID`: unique row locality id (Numeric)
- `LocalityName`: locality name (Numeric)
- `NearbyLocalityID`: Unique row locality id (Numeric)

**School data**
- `SchoolID`: unique row school id (Numeric)
- `SchoolName`: school name
- `LocalityID`: unique row locality id (Numeric)

**ShoppingCenter**
- `StoreID`: unique row store id (Numeric)
- `StoreName`: store name
- `Type`: type of shopping center (categorical: 'Grocery Store','Shopping Mall')
- `LocalityID`: unique row school id (Numeric)


<p style = 'font-size:16px;font-family:Arial'><b>Dataset source:</b> <a href = 'https://www.kaggle.com/datasets/thedevastator/australian-housing-data-1000-properties-sampled'>kaggle</a></p>


<p style = 'font-size:16px;font-family:Arial'><b>Links:</b></p>
<ul style = 'font-size:16px;font-family:Arial'>
    <li>Teradataml Python reference: <a href = 'https://docs.teradata.com/search/all?query=Python+Package+User+Guide&content-lang=en-US'>here</a></li>
    <li>Langchain Python reference: <a href='https://python.langchain.com/docs/get_started/introduction.html'>here</a></li>
</ul>

<footer style="padding-bottom:35px; border-bottom:3px solid #91A0Ab">
    <div style="float:left;margin-top:14px">ClearScape Analytics™</div>
    <div style="float:right;">
        <div style="float:left; margin-top:14px">
            Copyright © Teradata Corporation - 2023, 2024. All Rights Reserved
        </div>
    </div>
</footer>